Mathlib/Topology/Compactness/CompactlyGeneratedSpace.lean

```diff
@@ -0,0 +1,177 @@
+if and only if for all compact Hausdorff space `K` and `f : K → X` continuous, `f ⁻¹' s` is closed
+in `K`. The typeclass `CompactlyGeneratedSpace` is defined by taking `K : Type u` when `X : Type u`,
+contrary to `UCompactlyGeneratedSpace` where spaces `K` can be taken in another universe.
+
+We prov basic properties and instances, and prove that a `SequentialSpace` is compactly generated,
+as well as Hausdorff `WeaklyLocallyCompactSpace`.
+
+## Main definitions
+
+* `CompactlyGeneratedSpace X`: the topology of `X` is coinduced by continuous maps coming from
+  compact Hausdorff spaces.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Compactly_generated_space>
+* <https://ncatlab.org/nlab/files/StricklandCGHWSpaces.pdf>
+
+## Tags
+
+compactly generated space
+-/
+
+universe u v
+
+open TopologicalSpace Filter Topology Set
+
+variable {X : Type u} {Y : Type v} [TopologicalSpace X] [TopologicalSpace Y]
+
+/-- A topological space `X : Type u` is compactly generated if its topology is coinduced by
+continuous maps `f : K → X`, where `K : Type u` is compact Hausdorff. -/
+abbrev CompactlyGeneratedSpace (X : Type u) [TopologicalSpace X] : Prop :=
+  UCompactlyGeneratedSpace.{u} X
+
+/-- A topological space `X` is compactly generated if a set `s` is closed when `f ⁻¹' s` is
+closed for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem compactlyGeneratedSpace_of_isClosed
+    (h : ∀ (s : Set X), (∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsClosed (f ⁻¹' s)) → IsClosed s) :
+    CompactlyGeneratedSpace X := by
+  refine uCompactlyGeneratedSpace_of_continuous_maps fun f h' ↦
+    continuous_iff_isClosed.2 fun t ht ↦ h _ ?_
+  intro K _ _ _ g hg
+  exact ht.preimage (h' (CompHaus.of K) { toFun := g, continuous_toFun := hg })
+
+/-- A topological space `X` is compactly generated if a set `s` is open when `f ⁻¹' s` is
+open for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem compactlyGeneratedSpace_of_isOpen
+    (h : ∀ (s : Set X), (∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsOpen (f ⁻¹' s)) → IsOpen s) :
+    CompactlyGeneratedSpace X := by
+  refine compactlyGeneratedSpace_of_isClosed fun s h' ↦ isOpen_compl_iff.1 <| h _ fun f hf ↦ ?_
+  rw [preimage_compl, isOpen_compl_iff]
+  exact h' f hf
+
+/-- In a compactly generated space `X`, a set `s` is closed when `f ⁻¹' s` is
+closed for every continuous map `f : K → X`, where `K` is compact Hausdorff. -/
+theorem CompactlyGeneratedSpace.isClosed' [CompactlyGeneratedSpace X] {s : Set X}
+    (hs : ∀ {K : Type u} [TopologicalSpace K], [CompactSpace K] → [T2Space K] →
+      ∀ (f : K → X), Continuous f → IsClosed (f ⁻¹' s)) : IsClosed s := by
+  rw [eq_compactlyGenerated (X := X), TopologicalSpace.compactlyGenerated, isClosed_coinduced,
```

Comment:
This theorem should hold for more general universes and `UCompactlyGeneratedSpace` as well. Could you try to prove it in the proper generality? If it's annoying to apply it (e.g. if Lean can't figure out the universes automatically in the proof of the theorem below), then you can state this special case as well.